File: PassInterim/passinterim.js
// NE PAS TOUCHER
titans = ['16;8;1155', '12;10;1081', '3;10;270', '17;7;1073', '13;9;1059', '22;6;1192'];
habitations = ['21;36', '26;23', '16;48'];
gaz = 1112;
// NE PAS TOUCHER

const parseTitan = (titanStr) => {
    const [heightStr, speedStr, pvStr] = titanStr.split(';');
    return {
        height: parseInt(heightStr),
        speed: parseInt(speedStr),
        pv: parseInt(pvStr)
    };
};

const parseHabitation = (habitationStr) => {
    const [heightStr, distanceStr] = habitationStr.split(';');
    return {
        height: parseInt(heightStr),
        distance: parseInt(distanceStr)
    };
};

const parsedTitans = titans.map(parseTitan);
const parsedHabitations = habitations.map(parseHabitation);
gaz = parseInt(gaz);

parsedTitans.sort((a, b) => b.height - a.height);

parsedHabitations.sort((a, b) => a.height - b.height);

let totalPoints = 0;
let currentGaz = gaz;
let currentHeight = 0;

const calcPowerAndGaz = (height, titanHeight, distance, speed, above) => {
    let power, gazConsumed;
    if (above) {
        power = (height - titanHeight) * 10 + distance * 2 - speed;
        gazConsumed = (height - titanHeight + distance);
    } else {
        power = Math.abs(height - titanHeight) * 5 + distance * 2 - speed;
        gazConsumed = Math.abs(height - titanHeight + distance);
    }
    return { power, gazConsumed };
};

for (const titan of parsedTitans) {
    let { height: tHeight, speed: tSpeed, pv: tPv } = titan;
    let bestHabitation = null;
    let above = false;

    for (const habitation of parsedHabitations) {
        const { height: hHeight, distance: hDistance } = habitation;
        if (hHeight > tHeight) {
            bestHabitation = habitation;
            above = true;
            break;
        } else {
            bestHabitation = habitation;
        }
    }

    if (bestHabitation === null) continue;

    const { height: hHeight, distance: hDistance } = bestHabitation;

    const moveGaz = Math.abs(currentHeight - hHeight);
    if (currentGaz < moveGaz) break;

    currentGaz -= moveGaz;
    currentHeight = hHeight;

    while (tPv > 0) {
        const { power, gazConsumed } = calcPowerAndGaz(hHeight, tHeight, hDistance, tSpeed, above);
        if (currentGaz < gazConsumed) break;

        tPv -= power;
        totalPoints += 1; // 1 pt par coup d'épée
        currentGaz -= gazConsumed;
    }

    if (tPv <= 0) {
        totalPoints += 100; // 100 pts par titan abattu
    }
}

console.log(totalPoints);

File: PassInterim/passinterim.py
# NE PAS TOUCHER
titans = ['16;8;1155', '12;10;1081', '3;10;270', '17;7;1073', '13;9;1059', '22;6;1192']
habitations = ['21;36', '26;23', '16;48']
gaz = 1112
# NE PAS TOUCHER

titans = [tuple(map(int, t.split(';'))) for t in titans]
habitations = [tuple(map(int, h.split(';'))) for h in habitations]
gaz = int(gaz)
titans.sort(key=lambda x: x[0], reverse=True)

habitations.sort(key=lambda x: x[0])

total_points = 0
current_gaz = gaz
current_height = 0

def calc_power_and_gaz(height, titan_height, distance, speed, above):
    if above:
        power = (height - titan_height) * 10 + distance * 2 - speed
        gaz_consumed = (height - titan_height + distance)
    else:
        power = abs(height - titan_height) * 5 + distance * 2 - speed
        gaz_consumed = abs(height - titan_height + distance)
    return power, gaz_consumed

for t_height, t_speed, t_pv in titans:
    best_habitation = None
    above = False

    for h_height, h_distance in habitations:
        if h_height > t_height:
            best_habitation = (h_height, h_distance)
            above = True
            break
        else:
            best_habitation = (h_height, h_distance)

    if best_habitation is None:
        continue

    h_height, h_distance = best_habitation

    move_gaz = abs(current_height - h_height)
    if current_gaz < move_gaz:
        break

    current_gaz -= move_gaz
    current_height = h_height

    while t_pv > 0:
        power, gaz_consumed = calc_power_and_gaz(h_height, t_height, h_distance, t_speed, above)
        if current_gaz < gaz_consumed:
            break

        t_pv -= power
        total_points += 1  # 1 pt par coup d'épée
        current_gaz -= gaz_consumed

    if t_pv <= 0:
        total_points += 100  # 100 pts par titan abattu

print(total_points)
